### ISSUE TITLE
Auto-generate liturgical colors

### ISSUE THREAD

**Issue by jarosz** [romcal/romcal#121]
I've noticed some optional memorials (even in general.js) have metadata with liturgical colors or titles ("saintsCyrilMonkAndMethodiusBishop"), and some don't (e.g. "theMostHolyNameOfJesus").

Some have only title without color (e.g. "saintPeterDamianBishopAndDoctorOfTheChurch").

Should optional memorials have this metadata at all? It's problematic for coincident memorials, like "saintSixtusIiPopeAndCompanionsMartyrsSaintCajetanPriest".

Regardless of the answer I think liturgical color could be auto-generated for most common combinations of data (WHITE on feast and memorial, RED if there's martyr title, there could be an APOSTLE title for red color as well, and so on).

**Comment by tukusejssirs** [comment]
I agree with you that it might be automated (not only for optional memorials, but for any and all celebrations) and if needed, the calendarists could override it when defining the particular celebration.

On the other hand, I have no idea if we should deal with this right away, as for `v2.0.0` we plan to separate the parts of the celebration names (e.g., translate all saints per person and add some metadata to them, like the standard titles, is he/she a saint or blessed, etc) and use an i18n tool; and then we could automate this (using the saint/celebration metadata for liturgical colour).

What do you think, @emagnier?

**Comment by jarosz** [comment]
I'll leave optional memorials for chinese region with no metadata for now (#118).

**Comment by emagnier** [comment]
Yes it could be automated and it could be handful. But in another way, I like to see all the metadata defined at the same place when reading a calendar file. And I'm sure that exceptions or rules exist on some celebrations, that might difficult to catch if we don't check the final data computed by romcal. For now I'm not too convinced to add this automation.

However the coincident memorials are effectively an issue. I'm working on that currently, and it's not an easy one since it's deeply wired in romcal.

**Comment by jarosz** [comment]
Rules given in the General Instruction of th Roman Missal are explicit and I don't know about any exceptions.

Conferences of Bishops may introduce some adaptations regarding color of the liturgical vestments (point 390), but they are usually generic, like "Gold- or silver-colored vestments may be worn on more solemn occasions in the dioceses of the United States of America" (point 346 h., not present in Latin version).

**Comment by emagnier** [comment]
Thanks for your message and the reference to the General Missal Instructions.
Yes on a second thought I think this will be a good addition 👍 
Rules from the point 346 are effectively very explicit. And we could always adjust a color case by case if we found exceptions.

**Comment by tukusejssirs** [comment]
@emagnier, just to be sure: in the TS branch, can we safely remove all liturgical colours from celebration definitions?

**Comment by emagnier** [comment]
Romcal try to determine the right `liturgicalColor`, if no one has been provided.
Note that all weekdays/sundays generated by romcal already have their right liturgical colors. So this new feature only applies to new liturgical days defined in the general or any particular calendars.

Here are the checks:

### 1. A liturgical color(s) has already been defined:

Nothing more to do: returns the already defined color(s).

### 2. No liturgical color has been defined:

Try to find the right default color...

1. If the liturgical day isn't a `COMMEMORATION` or a `WEEKDAY`
    - and is for a `MARTYR`: return **RED**.
    - and is not for a `MARTYR`: return **WHITE**.

2. Then, if the liturgical day is a `COMMEMORATION` or a `WEEKDAY`, and is celebrated during `LENT` or `ADVENT`: return **PURPLE**.

3. Then, if the liturgical day is a `COMMEMORATION` or a `WEEKDAY`, and is celebrated during `ORDINARY_TIME`: return **GREEN**.

4. Otherwise, return **WHITE** that match all other seasons, or as a default color.

---

As you see, all cases might not be covered. The test for Apostles is missing (maybe we should add it?), but there is still exceptions: St-John Apostle on December 27 is WHITE.

In the general/particular calendars, I guess we could remove the defined liturgical colors for all common cases, and just keep them on exceptions (like Saint John).

**Comment by tukusejssirs** [comment]
I think we should add all known stuff:
- apostles;
- Chair of St Peter;
- Conversion of St Paul;
- anything else we become aware of.

I suggest to manually check tbe already defined colours and remove those which are correctly generated. The rest (i.e. those that are’t) should be used to improve the ‘generator’ function. In the end, we should not have a colour def in the celeb def (but it should be still possible to set it there).

**Comment by emagnier** [comment]
This function is just a helper to address missing colours, with large criteria, not to manage exceptions.

If you want to manage exceptions or specify a different colour only on one of the items, the general/particular calendar files are still the best place. I realise that I forgot the apostles and I can add this criteria, but that's all.

**Comment by tukusejssirs** [comment]
> This function is just a helper to address missing colours, with large criteria, not to manage exceptions.

Is there any reason why you are against making this function a function to ‘manage exceptions’ / manage liturgical colours for all celebrations according to the the GIRM rules? I thought that was the primary reason why @jarosz opened this issue.

Anyway, if you really don’t want to add anything else but the explicit rules from the GIRM, you need to check these rules (note: I didn’t check if they are already implemented):
- `#ffffff` [white]:
   - Eastertide;
   - Christmastide;
   - celebrations of the Lord other than of his Passion;
   - celebrations of the Blessed Virgin Mary;
   - celebrations of the Holy Angels;
   - celebrations of saints who were not martyrs;
   - All Saints;
   - Nativity of St John the Baptist;
   - St John the Evangelist;
   - Chair of St Peter;
   - Conversion of St Paul;
- `#ff0000` [red]:
   - Palm Sunday of the Lord’s Passion;
   - Good Friday;
   - Pentecost Sunday;
   - celebrations of the Lord’s Passion;
   - feasts of the apostles;
   - feasts of the evangelists;
   - celebrations of martyr saints;
- `#00ff00` [green]:
   - Ordinary Time;
- `#6600cc` [violet/purple];
   - Advent;
   - Lent;
   - offices and Masses for the dead;
- `#ff5ccd` [pink besides violet/purple]:
   - Third Sunday of Advent [Gaudete Sunday];
   - Fourth Sunday of Lent [Lætare Sunday];
- `#000000` [black besides purple]:
   - All Souls;
   - offices and Masses for the dead.

Note that I have removed all the US rules and based the rules on the official Latin GIRM from _editio typica tertia_ from 2002.

**Comment by emagnier** [comment]
> Is there any reason why you are against making this function a function to ‘manage exceptions’ 

Because the right place to manage exceptions and refine metadata for any special days are the calendar files itself, where are defined the liturgical days. It doesn't make sense to create another way to define liturgical colors for specific days. Let's keep things simple with the tools we already have, and don't complexity the data architecture. Our library architecture is already able to manage it. This was already my point of view in a [previous comment](https://github.com/romcal/romcal/issues/121#issuecomment-565817958) and you added a thumbs up. 

Also please keep a bit of humility in your comments. No need to argue over and over again when you don't agree. The primary reason for this issue was to automate things when it's easily predictable, not to manage exceptions.

**Comment by tukusejssirs** [comment]
> Also please keep a bit of humility in your comments. No need to argue over and over again when you don't agree. The primary reason for this issue was to automate things when it's easily predictable, not to manage exceptions.

I am awfully sorry I touched your feelings. :cry: I didn’t mean it.

> > Is there any reason why you are against making this function a function to ‘manage exceptions’
> 
> Because the right place to manage exceptions and refine metadata for any special days are the calendar files itself, where are defined the liturgical days. It doesn't make sense to create another way to define liturgical colors for specific days. Let's keep things simple with the tools we already have, and don't complexity the data architecture. Our library architecture is already able to manage it. This was already my point of view in a [previous comment](https://github.com/romcal/romcal/issues/121#issuecomment-565817958) and you added a thumbs up.

Indeed. Okay, I agree that the function should not manage exceptions.

However, as I pointed out in my previous comment (which was rude), you did not implement the whole point 346. I know that most of it is already done, however, not all of those rules.

**Comment by emagnier** [comment]
You are a bit stubborn... the point 346 is already implemented in romcal (thanks to calendar definitions that manage specifics for special days, especially in the celebrations/general.ts files). You seem to still not understand, but I'll not add complexity to the romcal architecture and not add extra logic here, when romcal is already able to manage celebration specifics very efficiently (including the point 346). The romcal architecture and codebase should always stay simple and easy to maintain as far as we can.

For the other particular calendars, you will probably never need to set the liturgical color property, thanks to this new feature, which is the primary goal of this issue.

**Comment by tukusejssirs** [comment]
Okay, @emagnier, I won’t comment on this anymore. In the end, you’re the dev, not me. My opinion didn’t change though, however, I’ll accept your decision and thus this issue will be closed once #250 is merged and when v2.0.0 is released.

---

> You are a bit stubborn...

I am sorry, I am **too** stubborn, because I’d like to help romcal to be as perfect as possible. However, I waste much of your time, for which I am very, very much sorry.

**Comment by emagnier** [comment]
It's good to want the best for romcal 👍. But indeed we also need to take in consideration all the technical points, which can have an incident on performance, codebase maintainability, and finally on contributor experience. The challenge is to find a good balance between all of that and feature design, without compromise data results. This is also why it's important to stay open minded to whatever others can share (experience, needs, view...). Be part of an open-source project like romcal is also a challenge of humility, patience towards others, and listening (I'm including me of course), but it's a good one! 😎

**Comment by tukusejssirs** [comment]
I agree with you.

> we also need to take in consideration all the technical points, which can have an incident on performance, codebase maintainability, and finally on contributor experience. The challenge is to find a good balance between all of that and feature design, without compromise data results.
>
> This is also why it's important to stay open minded to whatever others can share (experience, needs, view...).
>
> Be part of an open-source project like romcal is also a challenge of humility, patience towards others, and listening (I'm including me of course), but it's a good one! :sunglasses:

This is why I alway address you my comments. I know for sure that I more often slow you down than give constructive (user) input, however, I think that if I stop giving you the feedback, you might miss the constructive part of it. Please bear with me and my stubbornness and try to look at my feedback/comments from the bird’s view. Please don’t let my commets touch your feelings; it is never my intention. Thank you very much in advance. :smiley:

**Comment by emagnier** [comment]
No worries, your comments didn't hurt me. Thanks for your message. 😃

**Comment by emagnier** [comment]
Addressed on https://github.com/romcal/romcal/pull/250.
